### PR TITLE
Feature: Open folder in new tab when dragging folder to "add tab" button

### LINKED ISSUE #4710: Feature: Drag folder onto "+" icon to create new tab
**Is your feature request related to a problem?**
It would be nice to be able to drag a folder onto the "+" icon for creating a new tab or onto a blank spot on the tab strip to create a new tab. Chrome/Edge/Firefox support doing this with links, and I have gotten used to doing it in browsers. When you try in Edge, a little arrow pops up too, which is a nice touch!
image

**Describe the solution**
Implementing something like I described above would be nice!

**Describe alternatives you've considered**
CTRL+click on a folder to do this would work too, but both would be even better!

**Additional comment**
Thank you for the work y'all are putting into this app, it's amazing!

### PR DESCRIPTION
**Pull request by denisdekh**
**Resolved / Related Issues**
- [x] Were these changes approved in an issue or discussion with the project maintainers? 
   Closes #4710 

Also added support for drag opening archives, but commented it out for now as the best solution I was able to find was too ugly in my opinion (parsing the extention from the filename and checking it agaist a list of zip extentions)

**Validation**
How did you test these changes?
- [x] Did you build the app and test your changes?
- [ ] Did you check for accessibility? You can use Accessibility Insights for this.
- [ ] Did you remove any strings from the en-us resource file?
   - [ ] Did you search the solution to see if the string is still being used? 
- [ ] Did you implement any design changes to an existing feature?
   - [ ] Was this change approved?
- [x] Are there any other steps that were used to validate these changes?
1. Ran the app
2. Tested dragging a single folder / multiple folders
3. Works even if non-folder items are selected and dragged (those are ignored)
4. Stress tested to some extend with dragging 50+ folders at once

**Screenshots (optional)**
Too hard to take a screenshot while dragging lol

### CODE REVIEW

**Comment by yaira2** [comment]
@denisdekh you can fix the error by running the XAML Styler extension on that file.

File: src/Files.App/Views/MainPage.xaml.cs [hunk 1]
```diff
@@ -487,6 +489,49 @@ private void RootGrid_PreviewKeyDown(object sender, KeyRoutedEventArgs e)
...
+
+			var deferral = e.GetDeferral();
+			try {
+				foreach (var item in items)
+					await NavigationHelpers.OpenPathInNewTab(item.Path);
+				deferral.Complete(); 
+			} catch { } 
+			lockFlag = false;
+		}
+		private async void HorizontalMultitaskingControlAddButton_DragOver(object sender, DragEventArgs e)
```
Comment (yaira2) [inline src/Files.App/Views/MainPage.xaml.cs:513]:
```suggestion

		private async void HorizontalMultitaskingControlAddButton_DragOver(object sender, DragEventArgs e)
```

File: src/Files.App/Views/MainPage.xaml.cs [hunk 2]
```diff
@@ -487,6 +489,49 @@ private void RootGrid_PreviewKeyDown(object sender, KeyRoutedEventArgs e)
...
+		{
+			if (!FilesystemHelpers.HasDraggedStorageItems(e.DataView)) { e.AcceptedOperation = DataPackageOperation.None; return; }
```
Comment (yaira2) [inline src/Files.App/Views/MainPage.xaml.cs:515]:
```suggestion
			if (!FilesystemHelpers.HasDraggedStorageItems(e.DataView))
			{
				e.AcceptedOperation = DataPackageOperation.None; return;
			}
```

File: src/Files.App/Views/MainPage.xaml.cs [hunk 3]
```diff
@@ -487,6 +489,54 @@ private void RootGrid_PreviewKeyDown(object sender, KeyRoutedEventArgs e)
...
+				return;
+			}
+
+			bool hasValidDraggedItems = 
+				(await FilesystemHelpers.GetDraggedStorageItems(e.DataView)).Any(x => x.ItemType is FilesystemItemType.Directory
+				//|| dropableArchiveTypes.Contains(x.Name.Split('.').Last().ToLower())
+				);
+
+			if (!hasValidDraggedItems) { e.AcceptedOperation = DataPackageOperation.None; return; }
+
```
Comment (yaira2) [inline src/Files.App/Views/MainPage.xaml.cs:528]: Can you split the code onto multiple lines to match the conventions used in the rest of the code base?

Comment (denisdekh) [reply src/Files.App/Views/MainPage.xaml.cs:528]: Thought it got fixed when I pressed commit suggestion, oops

**Comment by denisdekh** [comment]
> @denisdekh you can fix the error by running the XAML Styler extension on that file.

Thanks for the tip, didn't know about the extention

File: src/Files.App/Views/MainPage.xaml.cs [hunk 4]
```diff
@@ -63,7 +66,6 @@ public MainPage()
 				FlowDirection = FlowDirection.RightToLeft;
 
 			UserSettingsService.OnSettingChangedEvent += UserSettingsService_OnSettingChangedEvent;
```
Comment (yaira2) [inline src/Files.App/Views/MainPage.xaml.cs:68]:
```suggestion
			UserSettingsService.OnSettingChangedEvent += UserSettingsService_OnSettingChangedEvent;

```

**Comment by yaira2** [comment]
This looks good behavior wise

**Comment by denisdekh** [comment]
Let me know if anything else needs to be done, I'll probably start looking for another issue to work on - suggestions are welcome. I am still not that familiar with app so it's hard to understand specifics at a glance

**Review by yaira2 — DISMISSED** [review]

File: src/Files.App/Views/MainPage.xaml.cs [hunk 5]
```diff
@@ -487,6 +490,58 @@ private void RootGrid_PreviewKeyDown(object sender, KeyRoutedEventArgs e)
...
+
+			try
+			{
+				e.Handled = true;
+				var deferral = e.GetDeferral();
+				e.DragUIOverride.IsCaptionVisible = true;
+				e.DragUIOverride.Caption = string.Format("OpenInNewTab".GetLocalizedResource());
+				e.AcceptedOperation = DataPackageOperation.Link;
+				deferral.Complete();
+			}  catch { }
```
Comment (0x5bfa) [inline src/Files.App/Views/MainPage.xaml.cs:542]: Insert a line break right after a closing curly bracket, according to the our codebase guideline.
```suggestion
			}
			catch { }
```

File: src/Files.App/Views/MainPage.xaml.cs [hunk 6]
```diff
@@ -487,6 +490,58 @@ private void RootGrid_PreviewKeyDown(object sender, KeyRoutedEventArgs e)
...
+				return;
+			}
+
+			bool hasValidDraggedItems = 
+				(await FilesystemHelpers.GetDraggedStorageItems(e.DataView)).Any(x => x.ItemType is FilesystemItemType.Directory
+				//|| dropableArchiveTypes.Contains(x.Name.Split('.').Last().ToLower())
+				);
+
+			if (!hasValidDraggedItems) 
+		   {
```
Comment (0x5bfa) [inline src/Files.App/Views/MainPage.xaml.cs:529]:
```suggestion
			{
```

File: src/Files.App/Views/MainPage.xaml.cs [hunk 7]
```diff
@@ -487,6 +490,58 @@ private void RootGrid_PreviewKeyDown(object sender, KeyRoutedEventArgs e)
...
+				.Where(x => x.ItemType is FilesystemItemType.Directory
+				//|| dropableArchiveTypes.Contains(x.Name.Split('.').Last().ToLower())
+				);
+
+			var deferral = e.GetDeferral();
+			try {
+				foreach (var item in items)
+					await NavigationHelpers.OpenPathInNewTab(item.Path);
+				deferral.Complete(); 
+			} catch { }
```
Comment (0x5bfa) [inline src/Files.App/Views/MainPage.xaml.cs:511]: As above.
```suggestion
			}
			catch { } 
```

File: src/Files.App/Views/MainPage.xaml.cs [hunk 8]
```diff
@@ -487,6 +490,58 @@ private void RootGrid_PreviewKeyDown(object sender, KeyRoutedEventArgs e)
...
+				return;
+			lockFlag = true;
+
+			var items = (await FilesystemHelpers.GetDraggedStorageItems(e.DataView))
+				.Where(x => x.ItemType is FilesystemItemType.Directory
+				//|| dropableArchiveTypes.Contains(x.Name.Split('.').Last().ToLower())
+				);
+
+			var deferral = e.GetDeferral();
+			try {
```
Comment (0x5bfa) [inline src/Files.App/Views/MainPage.xaml.cs:507]: Similar to the above.
```suggestion
			try
			{
```

File: src/Files.App/Views/MainPage.xaml.cs [hunk 9]
```diff
@@ -487,6 +490,58 @@ private void RootGrid_PreviewKeyDown(object sender, KeyRoutedEventArgs e)
...
+				foreach (var item in items)
+					await NavigationHelpers.OpenPathInNewTab(item.Path);
+				deferral.Complete();
```
Comment (0x5bfa) [inline src/Files.App/Views/MainPage.xaml.cs:510]:
```suggestion

				deferral.Complete(); 
```

File: src/Files.App/Views/MainPage.xaml.cs [hunk 10]
```diff
@@ -487,6 +490,58 @@ private void RootGrid_PreviewKeyDown(object sender, KeyRoutedEventArgs e)
 			}
 		}
 
+		private bool lockFlag = false;
+		//private string[] dropableArchiveTypes = { "zip", "rar", "7z", "tar" };
+		private async void HorizontalMultitaskingControlAddButton_Drop(object sender, DragEventArgs e)
+		{
+			if (lockFlag || !FilesystemHelpers.HasDraggedStorageItems(e.DataView))
+				return;
+			lockFlag = true;
```
Comment (0x5bfa) [inline src/Files.App/Views/MainPage.xaml.cs:499]:
```suggestion

			lockFlag = true;
```

File: src/Files.App/Views/MainPage.xaml.cs [hunk 11]
```diff
@@ -487,6 +490,58 @@ private void RootGrid_PreviewKeyDown(object sender, KeyRoutedEventArgs e)
 			}
 		}
 
+		private bool lockFlag = false;
+		//private string[] dropableArchiveTypes = { "zip", "rar", "7z", "tar" };
+		private async void HorizontalMultitaskingControlAddButton_Drop(object sender, DragEventArgs e)
```
Comment (0x5bfa) [inline src/Files.App/Views/MainPage.xaml.cs:495]:
```suggestion

		private async void HorizontalMultitaskingControlAddButton_Drop(object sender, DragEventArgs e)
```

**Review by 0x5bfa — CHANGES_REQUESTED** [review]
There's not enough line breaks to keep the codebase looks clean.

Otherwise, LGTM.

**Pull request merged**